### ISSUE TITLE
Clarify state of SMTP smuggling attack prevention

### ISSUE THREAD

**Issue by link2xt** [haraka/Haraka#3268]
With the recent news of [SMTP Smuggling](https://sec-consult.com/blog/detail/smtp-smuggling-spoofing-e-mails-worldwide/) attack it would be nice to double-check that Haraka is not vulnerable and clarify it in the docs.

From the code it seems that the code treats bare LF as a valid line ending:
https://github.com/haraka/Haraka/blob/389fcd509685d86e1ac41fcf3885f69620106e7c/connection.js#L327

If `.<LF>` is encountered instead of `.<CR><LF>`, an error is thrown and user gets to https://haraka.github.io/barelf:
https://github.com/haraka/Haraka/blob/389fcd509685d86e1ac41fcf3885f69620106e7c/connection.js#L1575-L1597

Does it mean `<LF>.<CR><LF>` instead of proper `<CR><LF>.<CR><LF>` can terminate the message?

I expect that only `<CR><LF>.<CR><LF>` should terminate the message. And bare LF should not be acceptable anywhere and ideally terminate the connection even if encountered elsewhere.

I did not test as I don't run Haraka anywhere (yet).

Published tools https://github.com/The-Login/SMTP-Smuggling-Tools may be helpful for testing, but just a test running in CI should be enough.

**Comment by ansuz** [comment]
according to a comment by @analogic in [this discussion](https://github.com/haraka/Haraka/discussions/3261) Haraka is unaffected.

**Comment by analogic** [comment]
Tested 
- Anything other than "\n.\r\n" will send a merged (proper and exploit part) email. 
- \n.\r\n" ends the transaction and creates a new one - it has the same measures as the previous one.

It looks like this:
```
← ehlo check.smtpsmuggling.com\r\n
→ 250-mail.poste.io Hello ... is at your service\r\n
→ 250-PIPELINING\r\n
→ 250-8BITMIME\r\n
→ 250-SMTPUTF8\r\n
→ 250-SIZE 26214400\r\n
→ 250 STARTTLS\r\n

← mail FROM:<test@check.smtpsmuggling.com> size=640\r\n
→ 250 sender <test@check.smtpsmuggling.com> OK\r\n

← rcpt TO:<...>\r\n
→ 250 recipient <...> OK\r\n

← 'data\r\n
→ 354 go ahead, make my day\r\n
← From: test@check.smtpsmuggling.com\r\nTo: ...\r\nSubject: CHECK EMAIL (\'\\n.\\r\\n\')\r\nDate: Sat, 09 Mar 2024 15:47:52 -0000\r\nMessage-ID: <170999567225.325610.15295146922005860938@check.smtpsmuggling.com>\r\n\r\nTESTING \'\\n.\\r\\n\' as "fake" end-of-data sequence!\r\n\n.\r\n\r\nmail FROM:<smuggled@check.smtpsmuggling.com>\r\nrcpt TO:<...>\r\ndata\r\nFrom: smuggled@check.smtpsmuggling.com\r\nTo: ...\r\nSubject: SMUGGLED EMAIL (\'\\n.\\r\\n\')\r\nDate: Sat, 09 Mar 2024 15:47:52 -0000\r\nMessage-ID: <170999567225.325610.10282728037216385338@check.smtpsmuggling.com>\r\n\r\nSMUGGLING WORKS with \'\\n.\\r\\n\' as "fake" end-of-data sequence!\r\n.\r\n
→ 250 Message Queued (9A1D4D7E-B97A-429E-987B-72EE4FBA859F.1) (9A1D4D7E-B97A-429E-987B-72EE4FBA859F.1)\r\n

→ 500 Unrecognized command\r\n
→ 250 sender <smuggled@check.smtpsmuggling.com> OK\r\n
→ 250 recipient <...> OK\r\n
→ 354 go ahead, make my day\r\n
→ 250 Message Queued (9A1D4D7E-B97A-429E-987B-72EE4FBA859F.2) (9A1D4D7E-B97A-429E-987B-72EE4FBA859F.2)\r\n
```

This is not the case with Postfix, which only checks the first email. If the second and smuggled email would be problematic, Haraka would happily refuse to process it.

**Comment by ansuz** [comment]
thanks for investigating!

**Comment by analogic** [comment]
> creating new transaction is dangerous because it will accept smuggled email

The point of this attack is to send email that should not be sent under normal circumstances. E.g. send proper email AND smuggle email that will later be DKIM signed and sent to the outside world without the restrictions that the previous email has. Creating a new transaction in Haraka means that every hook and every attribute will be run and checked in the same way as the previous email - Postfix will just pre-check the first headers, accept the whole package and then split it into two separate payloads.

I really don't know why "\n.\r\n" is used (qmail inspired?), but I see no danger in leaving it that way. I would rather not change those core parts that might affect compatibility. There is an old saying in our country - don't fix a toilet that doesn't leak.

**Comment by timglong** [comment]
This issue was closed when 3.0.3 was the current release. [This presentation](https://www.usenix.org/system/files/usenixsecurity25-wang-chuhan.pdf) from USENIX Security '25 (August 2025) indicates 3.0.3 was vulnerable to SMTP smuggling on both the sending and receiving ends. Looking through the release notes since 3.0.3, I don't see anything that appears related to this apparent vulnerability; however, I'm not an email expert. Can someone chime in on this topic?

Edit: From the linked report:
> Among different types of payload, A1 (\n.\n) and A5 (\n.\r\n) are the most impactful.

These are the two payloads identified as receiving vulnerabilities for 3.0.3.

**Comment by msimerson** [comment]
Great accusations require great evidence, and the evidence is still absent. "Someone said something on the internet" isn't evidence that Haraka was ever vulnerable to SMTP smuggling. We have seen SMTP smuggling reports but none withstood initial scrutiny. Our source is open and freely accessible for everyone to inspect and review.
